File: backend/src/main/java/io/github/artemy/osipov/shop/controller/ErrorHandler.java
package io.github.artemy.osipov.shop.controller;

import io.github.artemy.osipov.shop.controller.dto.DError;
import io.github.artemy.osipov.shop.exception.EntityNotFoundException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.HashMap;
import java.util.Map;

@Slf4j
@RestControllerAdvice
public class ErrorHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public DError processEntityNotFound(EntityNotFoundException ex) {
        log.info("entity not found", ex);
        return new DError(ex.getMessage());
    }

    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(
            MethodArgumentNotValidException ex,
            HttpHeaders headers,
            HttpStatus status,
            WebRequest request) {
        log.info("invalid request", ex);
        Map<String, String> errors = new HashMap<>();
        ex.getBindingResult().getFieldErrors().forEach((error) -> errors.put(
                error.getField(),
                error.getDefaultMessage()
        ));
        var response = new DError();
        response.setInternalMessage(ex.getMessage());
        response.setDetails(errors);

        return handleExceptionInternal(ex, response, headers, HttpStatus.BAD_REQUEST, request);
    }
}

File: backend/src/main/kotlin/io/github/artemy/osipov/shop/controller/ErrorHandler.kt
package io.github.artemy.osipov.shop.controller

import io.github.artemy.osipov.shop.controller.dto.DError
import io.github.artemy.osipov.shop.exception.EntityNotFoundException
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

private val logger = KotlinLogging.logger {}

@RestControllerAdvice
class ErrorHandler() {

    @ExceptionHandler
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleEntityNotFound(ex: EntityNotFoundException): DError {
        logger.info(ex) { "entity not found" }
        return DError(internalMessage = ex.message ?: "")
    }

    @ExceptionHandler
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleMethodArgumentNotValid(ex: MethodArgumentNotValidException): DError {
        logger.info(ex) { "invalid request" }
        val errors: MutableMap<String, String> = mutableMapOf()
        ex.bindingResult.fieldErrors.forEach { error: FieldError ->
            errors[error.field] = error.defaultMessage ?: ""
        }
        return DError(
            internalMessage = ex.message ?: "",
            details = errors
        )
    }
}